### PR TITLE
[DEVOPS-720] nix 2.x upgrade for infrastructure

### PR DESCRIPTION
**Pull request by disassembler**
Upgrading infrastructure machines (hydra, deployer, buildkite) to nix 2.x.

### CODE REVIEW

File: deployments/infrastructure-target-aws.nix [hunk 1]
```diff
@@ -17,7 +17,9 @@ in rec {
       associatePublicIpAddress = true;
       securityGroups = [
         resources.ec2SecurityGroups."allow-deployer-ssh-${region}-${org}"
-        resources.ec2SecurityGroups."allow-public-www-${region}-${org}"
+        resources.ec2SecurityGroups."allow-hydra-ssh-${region}-${org}"
+        resources.ec2SecurityGroups."allow-public-www-https-${region}-${org}"
+        resources.ec2SecurityGroups."allow-public-www-http-${region}-${org}"
```
Comment (cleverca22) [inline deployments/infrastructure-target-aws.nix]: this fixed being unable to open http://hydra.iohk.io to get the redirect
and also hydra being unable to ssh its own slaves

File: modules/hydra-master.nix [hunk 2]
```diff
@@ -41,14 +41,17 @@ in {
...
+      (mkLinux nodes.hydra-build-slave-1.config.networking.publicIPv4)
+      (mkLinux nodes.hydra-build-slave-2.config.networking.publicIPv4)
       (mkMac "de302.macincloud.com")
       (mkMac "du516.macincloud.com")
       (mkMac "de528.macincloud.com")
     ];
-    extraOptions = "auto-optimise-store = true";
+    extraOptions = ''
+      auto-optimise-store = true
+      allowed-uris = https://github.com/NixOS/nixpkgs/archive
```
Comment (cleverca22) [inline modules/hydra-master.nix]: this allows the new hydra build to fetch nixpkgs

File: nixpkgs-src.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
     "owner":  "NixOS",
     "repo":   "nixpkgs",
-    "rev": "3aec59c99ff6692468a069fa8a8d6a05000fca81",
-    "sha256": "0z0scsbxiscj43j9zms3jkc3f7jj4vnijsjpv5z6smc4s34ddfxs"
+    "rev": "06c576b0525da85f2de86b3c13bb796d6a0c20f6",
+    "sha256": "0kphjh8cqf8kclirw740z9zlm8fl3zdyli878qyvnsi9zmc6cv1w"
```
Comment (cleverca22) [inline nixpkgs-src.json]: the tip of the nixos-18.03 branch at the time the PR was being written

File: modules/hydra-slave.nix [hunk 4]
```diff
@@ -7,11 +9,28 @@ in
   services = {
     dd-agent.tags              = [" group:hydra-and-slaves"];
   };
+  nixpkgs.overlays = [
+    (self: super: {
+      nix' = super.nix.overrideAttrs ( drv: {
+        patches = [
+          (pkgs.fetchpatch {
+             url = "https://github.com/NixOS/nix/commit/a6c5955c8d59ca26745843f9779253a173876665.patch";
```
Comment (deepfire) [inline modules/hydra-slave.nix]: I think it's worth mentioning why this is needed, as a comment -- so we have context on when to remove this overlay.

**Pull request merged**